gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/FlowSpec.java

```diff
@@ -423,6 +423,17 @@ public boolean isScheduled() {
     return getConfig().hasPath(ConfigurationKeys.JOB_SCHEDULE_KEY);
   }
 
+  /**
+   * Create a new FlowSpec object with the added property defined by path and value parameters
+   * @param path key for new property
+   * @param value
+   */
+  public FlowSpec addProperty(String path, String value) {
+    Properties properties = this.getConfigAsProperties();
+    properties.setProperty(path, value);
+    return new Builder(this.getUri()).withConfigAsProperties(properties).build();
```

Comment:
Do you intend to clone a new FlowSpec?

Reply:
This should create a new FlowSpec with the shared `configAsProperties` object. I think you did not mean to do that.

Reply:
also, i think we should add configs to `config` field also. that is a better source of truth.

Reply:
Yes I have to clone and create a new FlowSpec because `type safe Configs` are immutable plus we have made `Config` a final variable (All variables are final). Rather than change what may be a core idea that `FlowSpecs` are immutable, I am creating a clone which has all the previous properties and URI plus the new one.

Reply:
I can also update the `Config` field, I thought it would get sync'd with the Properties but doesn't look like it. Making a change and updating test too.

Reply:
My concern here is mostly that it would create a decent amount of object churn, given the size of configs it should lead to an increase of jvm GC given that this would happen once per execution.

That being said it might be an overoptimization to enforce such a rule until you instrument results on a high load, the alternative would be messy, need to add a way for compileSpec() to take in a flow execution ID and enforce using that ID instead of using the provided one in the flow config or from the system time, which would also lead to a bit of an anti-pattern of compiler expecting to use either the flow config's ID (that you provided here) or it's own generated ID.

Reply:
Yea, I do not prefer cloning too, but `config` is `final` here. Not sure if it should be.
Regarding the extra objects, the spec create in line 201 should lose reference by line 204, so we should be good at least in this case.

Reply:
with the name changed to `createFlowSpecWithProperty` it becomes more clear that this method is **creating** a new spec and callers should not hold the reference of the original spec. Maybe we can write it in the methods javadoc @umustafi